File: ios/Ripple/BLE/BleLink.swift
import Foundation

/// Common half of a BLE link: fragments outbound packets into frames and drains
/// them one at a time (waiting for the radio's "ready" signal between frames);
/// reassembles inbound frames into packets.
class BleLink: Link {
    let id: String
    var peerHex: String?
    /// Negotiated frame size. Updated by the subclass once the MTU is known.
    var frameSize: Int = 20

    // Diagnostics
    let openedAt = Date()
    private(set) var bytesIn = 0, bytesOut = 0, packetsIn = 0, packetsOut = 0
    var rssi: Int?
    private(set) var lastActivity = Date()
    let log = EventLog.global

    private let reassembler = Reassembler()
    private let workQueue = DispatchQueue(label: "app.ripple.mesh.link")
    private var pendingFrames: [Data] = []
    private var inFlight = false
    private(set) var isClosed = false
    let onPacket: (BleLink, Data) -> Void
    let onClosed: (BleLink) -> Void

    init(id: String, onPacket: @escaping (BleLink, Data) -> Void, onClosed: @escaping (BleLink) -> Void) {
        self.id = id; self.onPacket = onPacket; self.onClosed = onClosed
    }

    func send(_ packetBytes: Data) {
        workQueue.async { [self] in
            guard !isClosed else { return }
            pendingFrames.append(contentsOf: Fragmenter.fragment(packetBytes, frameSize: frameSize, streamId: UInt16.random(in: 0...UInt16.max)))
            pump()
        }
    }

    /// Subclass calls this when the radio can accept another frame.
    func radioReady() { workQueue.async { [self] in inFlight = false; pump() } }

    /// Subclass calls this for every frame received from the radio.
    func onFrame(_ frame: Data) {
        workQueue.async { [self] in
            bytesIn += frame.count; lastActivity = Date()
            if let packet = reassembler.push(frame) { packetsIn += 1; onPacket(self, packet) }
        }
    }

    private func pump() {
        guard !inFlight, !isClosed, !pendingFrames.isEmpty else { return }
        let frame = pendingFrames.removeFirst()
        inFlight = true
        // writeFrame returns false when the radio is busy; the frame goes back to the head
        // of the queue and we wait for radioReady().
        if writeFrame(frame) {
            bytesOut += frame.count; lastActivity = Date()
            if frame.count >= 4, Int(frame[frame.startIndex + 2]) + 1 == Int(frame[frame.startIndex + 3]) { packetsOut += 1 }
        } else {
            pendingFrames.insert(frame, at: 0)
        }
    }

    func close() {
        workQueue.async { [self] in
            guard !isClosed else { return }
            isClosed = true
            pendingFrames.removeAll()
            closeTransport()
            onClosed(self)
        }
    }

    /// Attempt to send one frame. Return true if accepted by the radio (a later
    /// radioReady() will release the next frame), false if the radio is busy.
    func writeFrame(_ frame: Data) -> Bool { fatalError("abstract") }
    func closeTransport() {}
}

File: android/app/src/main/java/app/ripple/mesh/ble/BleLink.kt
package app.ripple.mesh.ble

import app.ripple.mesh.core.Fragmenter
import app.ripple.mesh.core.Link
import app.ripple.mesh.core.Reassembler
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.random.Random

/**
 * Common half of a BLE link: fragments outbound packets into frames and pushes
 * them through a single writer thread; reassembles inbound frames into packets.
 *
 * Subclasses provide [writeFrame] (blocking until the frame is on the air) and
 * [closeTransport].
 */
abstract class BleLink(override val id: String, private val onPacket: (BleLink, ByteArray) -> Unit, private val onClosed: (BleLink) -> Unit) : Link {
    override var peerHex: String? = null

    /** Negotiated frame size (mtu - 3, capped at 512). Updated by the subclass. */
    @Volatile var frameSize: Int = 20

    private val reassembler = Reassembler()
    private val outbound = LinkedBlockingQueue<ByteArray>()
    private val closed = AtomicBoolean(false)
    private val writer = Thread({ writerLoop() }, "ble-writer-$id").apply { isDaemon = true }

    fun start() { writer.start() }

    override fun send(packetBytes: ByteArray) {
        if (!closed.get()) outbound.offer(packetBytes)
    }

    /** Called by the subclass for every frame received from the radio. */
    protected fun onFrame(frame: ByteArray) {
        val packet = synchronized(reassembler) { reassembler.push(frame) } ?: return
        onPacket(this, packet)
    }

    private fun writerLoop() {
        try {
            while (!closed.get()) {
                val packet = outbound.take()
                val streamId = Random.nextInt(0, 0x10000)
                for (frame in Fragmenter.fragment(packet, frameSize, streamId)) {
                    if (closed.get()) return
                    if (!writeFrame(frame)) { close(); return }
                }
            }
        } catch (_: InterruptedException) { }
    }

    override fun close() {
        if (!closed.compareAndSet(false, true)) return
        writer.interrupt()
        try { closeTransport() } catch (_: Exception) {}
        onClosed(this)
    }

    val isClosed: Boolean get() = closed.get()

    /** Write one frame and block until it has been sent. Return false on fatal error. */
    protected abstract fun writeFrame(frame: ByteArray): Boolean
    protected abstract fun closeTransport()
}
